### ISSUE TITLE
CSV.File Memory Leak

### ISSUE THREAD

**Issue by stugorf** [JuliaData/CSV.jl#368]
I am loading a local CSV file using CSV.File and setting types using the code below. The load works on files ~50MB with 125,000 rows. However, if I attempt to load a CSV that is ~1GB with significantly more rows, the Julia process memory exceeds 1GB and will rapidly grow to consume all of my workstations free memory.

I am using:
Julia v1.0.2
CSV v0.4.3
Missings v0.3.1
DataFrames v0.15.2

# Load CSV
csv_df = CSV.File("historical.csv", types=Dict(
                "a"     => Union{Missing,String},
                "b"     => Union{Missing,String},
                "c"     => Union{Missing,String},
                "d"     => Union{Missing,String},
                "e"     => Union{Missing,String},
                "f"     => Union{Missing,String},
                "g"     => Union{Missing,String},
                "h"     => Union{Missing,String},
                "i"     => Union{Missing,String},
                "j"     => Union{Missing,String},
                "k"     => Union{Missing,String},
                "l"     => Union{Missing,String},
                "m"     => Union{Missing,String},
                "n"     => Union{Missing,String},
                "o"     => Union{Missing,String},
                "p"     => Union{Missing,String},
                "q"     => Union{Missing,String},
                "r"     => Union{Missing,String},
                "s"     => Union{Missing,String},
                "t"     => Union{Missing,String},
                "u"     => Union{Missing,String},
                "v"     => Union{Missing,String},
                "w"     => Union{Missing,String},
                "x"     => Union{Missing,String},
                "y"     => Union{Missing,String},
                "z"     => Union{Missing,String},
                "aa"    => Union{Missing,String},
                "ab"    => Union{Missing,String},
                "ac"    => Union{Missing,String},
                "ad"    => Union{Missing,String},
                "ae"    => Union{Missing,String})) |> DataFrame

Thank you for any insights into this issue.

**Comment by stugorf** [comment]
Milan,

Unfortunately, all of my data is protected health information (PHI). However, the can describe the data by saying all 32 fields are strings, most fields have less than 100 characters, and one field has up to 250 characters. I have files that exceed 1 million rows in the format described.

I agree that reading in a 1GB file may require more than 1GB of memory, but the simple CSV.File code that I included will consume all existing memory on a workstation, or allocated to a container, and eventually fail with an OutOfMemory error. For example, on my 64GB machine the Julia process running this code starts at ~312MB, then jumps to ~700MB, then proceeds to multiple GB. I let it run just to see how far it would go but stopped at 30GB.

Regards,

-David

________________________________

**Comment by stugorf** [comment]
Hi Jacob,

I really enjoy using CSV, thank you for developing it. I am on MacOS, and have tested the same code in an Ubuntu container.

Regards,

-David

________________________________

**Comment by nalimilan** [comment]
OK. Could you try generating a random file which reproduces the problem? For example, something like this should produce 32 columns with `N` rows containing random strings of up to 250 characters:
```julia
DataFrame(string.(rand(1:big(10)^250, N, 32)))
```
You can save that using `CSV.write` and see whether you are able to load it.

What's the rough percentage of unique values in each column?

**Comment by stugorf** [comment]
UPDATE:
I iterated the creation of the data frame and had similar memory consumption issues on write and hangs/consumption on reads for `df = DataFrame(string.(rand(1:big(10)^250, 100000, 32)))` and `df = DataFrame(string.(rand(1:big(10)^250, 100000, 32)))` where the resulting files were 800MB and 80MB  respectively. Then I ran `df = DataFrame(string.(rand(1:big(10)^250, 1000, 32)))` which resulted in a 8MB file and the write process consumed reasonable memory, and the read succeeded in < 15sec.

**Comment by nalimilan** [comment]
> Interestingly, the resulting data frame fields were typed as Float64 even though in the CSV.File I passed a Dict with types all `Union{Missing, String}` When I created the data frame I checked types using `typeof(df[1,1])` and the result is `String`.

AFAICT that's because the column names you give don't match those of the data frame. Anyway better try with `String` rather than `Union{String,Missing}` for a start and see whether the problem also happens. You can do ` df = DataFrame(string.('a', rand(1:big(10)^250, N, 32)))` to ensure values are parsed as strings.

**Comment by stugorf** [comment]
Injecting the char `a` into the creation of the data frame helped. I can now create, write, and read  the resulting .csv where the rowcount = 1,000,000 and filesize is 8GB. I left the type Dict as `Union{Missing, String}`. Now I just need to figure out why my 5GB file with same rowcount, columncount is failing and consuming so much memory. Julia consumed 24GB to create the data frame using `df = DataFrame(string.('a', rand(1:big(10)^250, 1000000, 32)))`

**Comment by nalimilan** [comment]
Can you show a few rows of the original file (you can replace characters with random ones to mask the actual values)?

**Comment by stugorf** [comment]
@nalimilan I have resolved the problem. I updated the version of DataFrames that I was using `DataFrames v0.14.1 ⇒ v0.15.2` during attempts to get the 5GB CSV to load. After updating, and ensuring the proper project was activated, I was able to load the CSV using CSV.File and passing a Dict for types as Union{Missing, String}. I ran into a different issue where using the loaded DataFrame as a variable in a string " DF: $csv_df" was getting evaluated when being passed as an argument to a function that prints the DataFrame when a certain debug level is set as `verbose`.  I should have caught that sending the string probably results in evaluation of the CSV regardless of what the function ends up doing with it. I wrongly assumed it would be a reference only. 

Thank you for all of your assistance and insights!
